File: service/src/api/websocket/game-router.ts
import { timingSafeEqual } from "node:crypto";

import type { FastifyInstance } from "fastify";

import {
  inboundGameEventSchema,
  type InboundGameEvent,
  type OutboundGameEvent,
} from "@takes-a-village/shared";

import type { UserDatabase } from "../../db.js";
import type { Game } from "../../game/game.js";
import { ConnectionManager, type SocketConnection } from "./connection-manager.js";
import { processGameEvent, type GameEventMessenger } from "./game-events.js";

export interface GameRouterDependencies {
  games: Map<string, Game>;
  connections: ConnectionManager;
  database: UserDatabase;
  botSecret: string;
  botServiceUrl?: string;
}

export function sameSecret(supplied: string, expected: string): boolean {
  const left = Buffer.from(supplied);
  const right = Buffer.from(expected);
  return left.length === right.length && timingSafeEqual(left, right);
}

function send(socket: SocketConnection, packet: OutboundGameEvent): void {
  socket.send(JSON.stringify(packet));
}

export function createGameEventMessenger(connections: ConnectionManager): GameEventMessenger {
  return {
    broadcast(gameId, packet) {
      connections.broadcast(gameId, packet);
    },
    sendPersonal(gameId, userId, packet) {
      connections.sendPersonal(gameId, userId, packet);
    },
    broadcastStates(game) {
      for (const [playerId] of connections.entries(game.id)) {
        const state = game.getStateForPlayer(playerId);
        if (state) connections.sendPersonal(game.id, playerId, { event: "game_state", data: state });
      }
    },
  };
}

export function cleanupDisconnectedPlayer(
  gameId: string,
  userId: string,
  dependencies: Pick<GameRouterDependencies, "games" | "connections" | "botSecret" | "botServiceUrl">,
): void {
  const game = dependencies.games.get(gameId);
  if (!game || game.status !== "WAITING") return;
  game.removePlayer(userId);
  dependencies.connections.broadcast(gameId, {
    event: "room_update",
    data: { player_count: game.players.size },
  });
  if (userId.startsWith("bot_") && dependencies.botServiceUrl) {
    void fetch(`${dependencies.botServiceUrl}/api/spawn_bots`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        gameId,
        botCount: 1,
        botSecret: dependencies.botSecret,
        botModel: "genetic",
        baseGenome: null,
      }),
    }).catch(() => undefined);
  }
}

export function registerGameWebSocketRoute(
  app: FastifyInstance,
  dependencies: GameRouterDependencies,
): void {
  const messenger = createGameEventMessenger(dependencies.connections);

  app.get("/ws", { websocket: true }, (socket, request) => {
    let joinedGameId: string | undefined;
    let joinedUserId: string | undefined;

    socket.on("message", async (raw) => {
      let packet: unknown;
      try {
        packet = JSON.parse(raw.toString());
      } catch {
        send(socket, { event: "error", data: { message: "Invalid WebSocket packet." } });
        return;
      }
      if (!packet || typeof packet !== "object" || Array.isArray(packet)) {
        send(socket, { event: "error", data: { message: "Invalid WebSocket packet." } });
        return;
      }
      const parsed = inboundGameEventSchema.safeParse(packet);
      if (!parsed.success) {
        send(socket, { event: "error", data: { message: "Malformed WebSocket packet." } });
        return;
      }

      if (joinedGameId && joinedUserId && dependencies.connections.get(joinedGameId, joinedUserId) !== socket) {
        send(socket, { event: "error", data: { message: "WebSocket connection was replaced." } });
        socket.close(4001);
        return;
      }

      if (parsed.data.event === "join_room") {
        if (joinedGameId) {
          send(socket, { event: "error", data: { message: "WebSocket is already joined to a game." } });
          return;
        }
        const { data } = parsed.data;
        const game = dependencies.games.get(data.gameId);
        if (!game) {
          send(socket, { event: "error", data: { message: "Game not found." } });
          return;
        }
        const browserAuthenticated = request.cookies.user_session === data.userId
          && await dependencies.database.userExists(data.userId);
        const botAuthenticated = data.userId.startsWith("bot_")
          && typeof data.botSecret === "string"
          && sameSecret(data.botSecret, dependencies.botSecret)
          && game.players.has(data.userId);
        if (!browserAuthenticated && !botAuthenticated) {
          send(socket, { event: "error", data: { message: "WebSocket authentication failed." } });
          return;
        }
        if (game.status !== "WAITING" && !game.players.has(data.userId)) {
          send(socket, { event: "error", data: { message: "Player is not a member of this game." } });
          return;
        }

        joinedGameId = data.gameId;
        joinedUserId = data.userId;
        game.addPlayer(data.userId);
        dependencies.connections.connect(socket, data.gameId, data.userId);
        if (game.hostId === data.userId && !game.hostConnected) {
          game.hostConnected = true;
          messenger.broadcastStates(game);
        }
        dependencies.connections.sendPersonal(data.gameId, data.userId, {
          event: "chat_history",
          data: game.getPrivateChatHistory(data.userId),
        });
        const state = game.getStateForPlayer(data.userId);
        if (state) dependencies.connections.sendPersonal(data.gameId, data.userId, { event: "game_state", data: state });
        messenger.broadcast(data.gameId, { event: "room_update", data: { player_count: game.players.size } });
        if (game.training && game.status === "RUNNING") {
          messenger.broadcast(game.id, { event: "game_started", data: { day: 1 } });
          messenger.broadcastStates(game);
        }
        return;
      }

      if (!joinedGameId || !joinedUserId) return;
      const game = dependencies.games.get(joinedGameId);
      if (!game) return;
      processGameEvent(game, joinedUserId, parsed.data as Exclude<InboundGameEvent, { event: "join_room" }>, messenger);
    });

    socket.on("close", () => {
      if (!joinedGameId || !joinedUserId) return;
      if (dependencies.connections.disconnect(socket, joinedGameId, joinedUserId)) {
        cleanupDisconnectedPlayer(joinedGameId, joinedUserId, dependencies);
      }
    });
  });
}

File: service/api/websocket/game_router.py
import asyncio
import hmac
import os

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from service.api.websocket.game_events import process_game_event


AUTHENTICATION_ERROR = {
    "event": "error",
    "data": {"message": "WebSocket authentication failed."},
}


def _join_error(websocket, message):
    return websocket.send_json({"event": "error", "data": {"message": message}})


def _is_authenticated(websocket, payload, game, database):
    user_id = payload.get("userId")
    if not isinstance(user_id, str) or not user_id:
        return False
    if user_id.startswith("bot_"):
        supplied_secret = payload.get("botSecret", "")
        expected_secret = os.environ.get("BOT_SECRET")
        return (
            bool(expected_secret)
            and isinstance(supplied_secret, str)
            and hmac.compare_digest(supplied_secret, expected_secret or "")
            and user_id in game.players
        )
    session_id = websocket.cookies.get("user_session")
    return (
        session_id == user_id
        and database.user_exists(user_id)
    )


def create_router(registry, manager, database, bot_client=None):
    router = APIRouter()

    @router.websocket("/ws")
    async def websocket_endpoint(websocket: WebSocket):
        await websocket.accept()
        user_id = game_id = None
        try:
            while True:
                packet = await websocket.receive_json()
                if not isinstance(packet, dict):
                    await _join_error(websocket, "Invalid WebSocket packet.")
                    continue
                event, payload = packet.get("event"), packet.get("data", {})
                if not isinstance(event, str) or not isinstance(payload, dict):
                    await _join_error(websocket, "Malformed WebSocket packet.")
                    continue

                if (user_id is not None
                        and manager.active_connections.get(
                            game_id, {}).get(user_id) is not websocket):
                    await _join_error(websocket, "WebSocket connection was replaced.")
                    await websocket.close(code=4001)
                    return

                if event == "join_room":
                    if user_id is not None:
                        await _join_error(websocket, "WebSocket is already joined to a game.")
                        continue
                    candidate_user_id = payload.get("userId")
                    candidate_game_id = payload.get("gameId")
                    game = registry.get(candidate_game_id)
                    if not game:
                        await _join_error(websocket, "Game not found.")
                        continue
                    if not _is_authenticated(websocket, payload, game, database):
                        await websocket.send_json(AUTHENTICATION_ERROR)
                        continue
                    if (game.status != "WAITING"
                            and candidate_user_id not in game.players):
                        await _join_error(
                            websocket, "Player is not a member of this game.")
                        continue
                    user_id, game_id = candidate_user_id, candidate_game_id
                    game.add_player(user_id)
                    await manager.connect(websocket, game_id, user_id)
                    if game.host_id == user_id and not game.host_connected:
                        game.host_connected = True
                        await manager.broadcast_game_state(game_id, game)
                    await manager.send_personal_message(
                        {"event": "chat_history",
                         "data": game.get_private_chat_history(user_id)}, game_id, user_id)
                    await manager.send_game_state(game_id, game, user_id)
                    await manager.broadcast_to_game(
                        {"event": "room_update", "data": {"player_count": len(game.players)}},
                        game_id)
                    if game.training and game.status == "RUNNING":
                        await manager.broadcast_to_game(
                            {"event": "game_started", "data": {"day": 1}}, game_id)
                        await manager.broadcast_game_state(game_id, game)
                elif game_id and user_id:
                    game = registry.get(game_id)
                    if game:
                        await process_game_event(
                            event, payload, game_id, user_id, game, manager)
        except WebSocketDisconnect:
            if game_id and user_id:
                disconnected = manager.disconnect(websocket, game_id, user_id)
                game = registry.get(game_id)
                if disconnected and game and game.status == "WAITING":
                    game.remove_player(user_id)
                    if user_id.startswith("bot_") and bot_client:
                        asyncio.create_task(bot_client.spawn_bots(
                            game_id, 1, None, "genetic", timeout=5.0))
                    asyncio.create_task(manager.broadcast_to_game(
                        {"event": "room_update",
                         "data": {"player_count": len(game.players)}}, game_id))
        except Exception:
            if game_id and user_id:
                manager.disconnect(websocket, game_id, user_id)

    return router
